fesvr/sai.cc

```diff
@@ -0,0 +1,116 @@
+#include "sai.h"
+#include "configstring.h"
+
+#define NHARTS_MAX 16
+
+void sai_t::host_thread(void *arg)
+{
+  sai_t *sai = static_cast<sai_t*>(arg);
+  sai->run();
+
+  while (true)
+    sai->target->switch_to();
+}
+
+sai_t::sai_t(const std::vector<std::string>& args) : htif_t(args)
+{
+  target = context_t::current();
+  host.init(host_thread, this);
+}
+
+sai_t::~sai_t(void)
+{
+}
+
+// Interrupt each core to make it start executing
+void sai_t::reset()
+{
+  uint32_t one = 1;
+  addr_t ipis[NHARTS_MAX];
+  int ncores = get_ipi_addrs(ipis);
+
+  for (int i = 0; i < ncores; i++)
+    write_chunk(ipis[i], sizeof(uint32_t), &one);
+}
+
+void sai_t::push_addr(addr_t addr)
+{
+  for (int i = 0; i < SAI_ADDR_CHUNKS; i++) {
+    in_data.push_back(addr & 0xffffffff);
+    addr = addr >> 32;
+  }
+}
+
+void sai_t::push_len(size_t len)
+{
+  for (int i = 0; i < SAI_LEN_CHUNKS; i++) {
+    in_data.push_back(len & 0xffffffff);
+    len = len >> 32;
+  }
+}
+
+void sai_t::read_chunk(addr_t taddr, size_t nbytes, void* dst)
+{
+  uint32_t *result = static_cast<uint32_t*>(dst);
+  size_t len = nbytes / sizeof(uint32_t);
+
+  in_data.push_back(SAI_CMD_READ);
+  push_addr(taddr);
+  push_len(len - 1);
```

Comment:
why do you use len - 1? You using integer division above which will truncate so don't need to add 1 to guarantee that all nbytes get sent?